backend/papersfeed/utils/papers/utils.py

```diff
@@ -99,41 +103,56 @@ def select_paper_search(args):
     # Page Number
     page_number = 1 if constants.PAGE_NUMBER not in args else int(args[constants.PAGE_NUMBER])
 
-    paper_ids = []
-    # exploit arXiv
-    try:
-        start = SEARCH_COUNT * (page_number-1)
-        print("[arXiv API] searching ({}~{})".format(start, start+SEARCH_COUNT-1))
-        arxiv_url = "http://export.arxiv.org/api/query"
-        query = "?search_query=" + urllib.parse.quote(keyword) \
-            + "&start=" + str(start) + "&max_results=" + str(SEARCH_COUNT)
-        response = requests.get(arxiv_url + query)
-
-        if response.status_code == 200:
-            response_xml = response.text
-            response_dict = xmltodict.parse(response_xml)['feed']
-            if 'entry' in response_dict and response_dict['entry']:
-                paper_ids = __parse_and_save_arxiv_info(response_dict)
-            else: # if 'entry' doesn't exist or it's the end of results
-                print("[arXiv API] entries don't exist")
-        else:
-            print("[arXiv API] error code {}".format(response.status_code))
-    except requests.exceptions.RequestException as exception:
-        print(exception)
-
-    # if cannot get any results from arXiv
-    if not paper_ids:
-        print("[naive-search] Searching in PapersFeed DB")
+    paper_ids = {
+        'papersfeed': [], # only used when all external search fails
+        'arxiv': [],
+        'crossref': [],
+    }
+    is_finished_dict = {
+        'arxiv': False,
+        'crossref': False,
+    }
+
+    future_to_source = {}
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
You can refer to [here](https://docs.python.org/ko/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) for these codes.